### PR TITLE
Epilog logics

### PR DESCRIPTION
**Pull request by TT-happy-work**

### CODE REVIEW

File: offline_data_crop.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import os
+from core.utils import draw_gt_bbox
+
+###################################################################################
+## This script preorganizes the Taggings of Yolov3 baseline 2 format [f_data_in_path]
+## into a desired [patch_w, patch_h] cropped Tagging file [f_data_out_path] and images [write_image_path]
+## main dir of cropped outputs
+
+###################################################################################
+
```
Comment (eyalpe) [inline offline_data_crop.py]: same comments as for rare_category_crop.py:
1. script template and configurable params
1. /home/tamar/

File: scripts/show_bboxes.py [hunk 2]
```diff
@@ -15,18 +15,23 @@
 import numpy as np
 from PIL import Image
 
-ID = 1
+
 label_txt = "./data/recce_data.txt"
-image_info = open(label_txt).readlines()[ID].split()
+label_txt = '/home/tamar/Desktop/dbg_anchor'
```
Comment (eyalpe) [inline scripts/show_bboxes.py]: debug remains ????
1. /home/tamar
1. don't re-declare a variable (`label_txt`)

File: scripts/show_bboxes.py [hunk 3]
```diff
@@ -15,18 +15,23 @@
...
+                                    int(float(bbox[1]))),
+                                    (int(float(bbox[2])),
+                                  int(float(bbox[3]))), (255,0,0), 2)
 
-image = Image.fromarray(np.uint8(image))
-image.show()
+    image = Image.fromarray(np.uint8(image))
+    print(image_path.split('/')[-1])
+    image.show()
+    pass
```
Comment (eyalpe) [inline scripts/show_bboxes.py]: redundant
```suggestion
```

File: train.py [hunk 4]
```diff
@@ -138,7 +138,7 @@ def __init__(self):
             variables_to_restore = [var for var in self.net_var if var.name.split(':')[0] in ckpt_net_var]
             self.loader = tf.train.Saver(variables_to_restore)
             # -nadav_wp_pruning
-            self.saver = tf.train.Saver(tf.global_variables(), max_to_keep=10)
+            self.saver = tf.train.Saver(tf.global_variables(), max_to_keep=20)
```
Comment (eyalpe) [inline train.py]: no reason of modifying this line, consider providing max_to_keep as configuration parameter

File: mAP/main.py [hunk 5]
```diff
@@ -314,13 +326,30 @@ def draw_plot_func(dictionary, n_classes, window_title, plot_title, x_label, out
...
   #print(txt_file)
   file_id = txt_file.split(".txt",1)[0]
   file_id = os.path.basename(os.path.normpath(file_id))
+  if ROC_FLAG:
+    file_id = file_id + '_ROC'
   # check if there is a correspondent predicted objects file
-  if not os.path.exists('predicted/' + file_id + ".txt"):
-    error_msg = "Error. File not found: predicted/" +  file_id + ".txt\n"
+  if not os.path.exists(predicted_path + file_id + ".txt"):
+    error_msg = "Error. File not found: predicted/" + file_id + ".txt\n"
```
Comment (eyalpe) [inline mAP/main.py]:
```suggestion
    error_msg = "Error. File not found: %s/" + file_id + ".txt\n" %predicted_path
```

File: evaluate.py [hunk 6]
```diff
@@ -113,12 +124,37 @@ def evaluate(self):
...
+                            xmin, ymin, xmax, ymax = list(map(str, coor))
+                            bbox_mess_ROC = ' '.join([class_name, score, xmin, ymin, xmax, ymax]) + '\n'
+                            f_ROC.write(bbox_mess_ROC)
+                            print('\t' + str(bbox_mess_ROC).strip())
 
                 if self.write_image:
-                    image = utils.draw_bbox(image, bboxes_pr, show_label=self.show_label)
-                    cv2.imwrite(self.write_image_path+image_name, image)
+                    image = utils.draw_bbox(image*255, bboxes_pr, show_label=self.show_label)
+                    cv2.imwrite(self.write_image_path+'reg/'+image_name, image)
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply evaluate.py]: OK

File: kmeans.py [hunk 7]
```diff
@@ -74,6 +74,7 @@ def kmeans(boxes, k, dist=np.median,seed=1):
             clusters[cluster] = dist(boxes[nearest_clusters == cluster], axis=0)
         last_clusters = nearest_clusters
 
+
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply kmeans.py]: ok

File: kmeans.py [hunk 8]
```diff
@@ -83,19 +84,41 @@ def parse_anno(annotation_path):
...
+#             ss=s[i].split(',')
+#
+#             x_min, y_min, x_max, y_max = float(ss[0]), float(ss[1]), float(ss[2]), float(ss[3])
+#             width  = (x_max - x_min) / image_w
+#             height = (y_max - y_min) / image_h
+#             result.append([width, height])
+#     result = np.asarray(result)
+#     return result
+
+
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply kmeans.py]: ok

File: kmeans.py [hunk 9]
```diff
@@ -129,13 +152,15 @@ def plot_cluster_result(clusters,nearest_clusters,WithinClusterSumDist,wh,k):
...
         for i in range(args.cluster_num):
             width, height = clusters[i]
             f.writelines(str(width) + " " + str(height) + " ")
 
-    WithinClusterMeanDist = np.mean(distances[np.arange(distances.shape[0]),nearest_clusters])
-    plot_cluster_result(clusters, nearest_clusters, 1-WithinClusterMeanDist, anno_result, args.cluster_num)
-
-
-
+    a=1
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply kmeans.py]: ok

File: mAP/main.py [hunk 10]
```diff
@@ -13,6 +13,9 @@
 random.seed(0)
 np.random.seed(0)
 
+# is the class-specific confidence-threshold is going to be considered (True) or all predicions are going to be cosidered
+ROC_FLAG = False
+
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply mAP/main.py]: ok

File: mAP/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply mAP/main.py]: ok

File: mAP/main.py [hunk 11]
```diff
@@ -734,6 +768,42 @@ def draw_plot_func(dictionary, n_classes, window_title, plot_title, x_label, out
...
+"""
+  Plot the statics of FAR and PD
+  PD  = TP[cls]/GT[cls]
+  FAR = FP[cls]/pred[cls]
+"""
+if draw_plot:
+  plt.figure()
+  window_title = "PD and FAR"
+  output_path = results_files_path + "/PD and FAR.png"
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply mAP/main.py]: ok

File: mAP/main.py [hunk 12]
```diff
@@ -734,6 +768,42 @@ def draw_plot_func(dictionary, n_classes, window_title, plot_title, x_label, out
...
+    else:
+      clsPD = len(conf_T[cls])/ gt_counter_per_class[cls]
+      clsFAR = len(conf_F[cls]) / pred_counter_per_class[cls]
+    PD.append(clsPD)
+    FAR.append(clsFAR)
+    CLS.append(cls)
+  width = 0.35
+  p1 = plt.bar(np.arange(N), PD, width)
+  p2 = plt.bar(np.arange(N), FAR, width,
+               bottom=PD)
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply mAP/main.py]: ok

File: mAP/main.py [hunk 13]
```diff
@@ -734,6 +768,42 @@ def draw_plot_func(dictionary, n_classes, window_title, plot_title, x_label, out
...
+  plt.ylabel('Statistics')
+  plt.title('PD and FAR Statistics per Class')
+  plt.xticks(np.arange(N), CLS)
+  plt.yticks(np.arange(0, 120, 5))
+  plt.legend((p1[0], p2[0]), ('PD', 'FAR'))
+
+
+
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply mAP/main.py]: ok

File: kmeans.py [hunk 14]
```diff
@@ -118,8 +141,8 @@ def plot_cluster_result(clusters,nearest_clusters,WithinClusterSumDist,wh,k):
 
 if __name__ == '__main__':
     parser = argparse.ArgumentParser()
-    parser.add_argument("--dataset_txt", type=str, default="./raccoon_dataset/train.txt")
-    parser.add_argument("--anchors_txt", type=str, default="./data/raccoon_anchors.txt")
+    parser.add_argument("--dataset_txt", type=str, default="/home/tamar/DBs/Reccelite/CroppedDB/croppedImgs_1_2_3_4_5_Th06_reg_rare.txt")
+    parser.add_argument("--anchors_txt", type=str, default="/home/tamar/RecceLite_code_packages/yolo3_baseline2/data/anchors/anchors_1-5_cropped_keras.txt")
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply kmeans.py]: OK

File: core/config.py [hunk 15]
```diff
@@ -12,33 +12,37 @@
 __C.YOLO.EXP_DIR                = "recce_",
 __C.YOLO.ROOT_DIR               = "./Runs/",
 __C.YOLO.CLASSES                = "./data/classes/recce.names"
-__C.YOLO.ANCHORS                = "./data/anchors/recce_anchors_2.txt"
+__C.YOLO.ANCHORS                = "./data/anchors/anchors_1-5_cropped.txt"
+__C.YOLO.ANCHORS                = "./data/anchors/anchors_reg_normalize.txt"
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply core/config.py]: ok

[the remark this answers is not in the document]
Comment (TT-happy-work) [reply core/config.py]: ok

File: core/config.py [hunk 16]
```diff
@@ -12,33 +12,37 @@
...
 __C.YOLO.ORIGINAL_WEIGHT        = "./checkpoint/yolov3_coco.ckpt"
 __C.YOLO.DEMO_WEIGHT            = "./checkpoint/yolov3_coco_demo.ckpt"
-__C.YOLO.IMAGE_HANDLE           = 'scale'  # 'crop' or 'scale'
-__C.YOLO.DATA_FORMAT            = 'NCHW'  # 'NCHW' or 'NHWC' (N-batch size, C-channels, H-height, W-width)
+__C.YOLO.IMAGE_HANDLE           = 'crop'  # 'crop' or 'scale'
+__C.YOLO.EPILOG_LOGICS          = True
+__C.YOLO.DATA_FORMAT            = 'NHWC'  # 'NCHW'(=opt for RT) or 'NHWC'(=default) (N-batch size, C-channels, H-height, W-width)
+__C.YOLO.CONF_TH_FILE           = "./data/classes/recce.confidence_th.txt"
+
+
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply core/config.py]: ok

File: kmeans.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply kmeans.py]: ok

File: offline_dataAugmentation_rareCategory_crop.py [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+
+###################################################################################
+## This script preorganizes the Taggings of Yolov3 baseline 2 format [f_data_in_path]
+## into a desired [patch_w, patch_h] cropped Tagging file [f_data_out_path] and images [write_image_path]
+## main dir of cropped outputs
+
+write_image_path = '/home/tamar/DBs/Reccelite/CroppedDB/croppedImgs_1_2_3_4_5_Th06_rare/'
+
+# the Tagging to be cropped:
+f_data_in_path = '/home/tamar/RecceLite_code_packages/yolo3_baseline2/data/individual_datasets_without_24/recce_data_Tagging_1_2_3_4_5.txt'
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply offline_dataAugmentation_rareCategory_crop.py]: ok

File: offline_dataAugmentation_rareCategory_crop.py [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+
+# the outout of this script: the cropped Tagging
+f_data_out_path = '/home/tamar/DBs/Reccelite/CroppedDB/croppedImgs_1_2_3_4_5_Th06_rare.txt'
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply offline_dataAugmentation_rareCategory_crop.py]: ok

File: offline_dataAugmentation_rareCategory_crop.py [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+
+
+patch_w = 1*800
+patch_h = 1*640
+
+cropTH = 0.6 # above this percentage of the bbox included in the path - clip the box. else - ignore the bbox.
+###################################################################################
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply offline_dataAugmentation_rareCategory_crop.py]: ok

File: offline_dataAugmentation_rareCategory_crop.py [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+import os
+from core import utils
+from core.utils import draw_gt_bbox
+import math
+
+###################################################################################
+## This script preorganizes the Taggings of Yolov3 baseline 2 format [f_data_in_path]
+## into a desired [patch_w, patch_h] cropped Tagging file [f_data_out_path] and images [write_image_path]
+## main dir of cropped outputs
+
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply offline_dataAugmentation_rareCategory_crop.py]: ok

File: offline_dataAugmentation_rareCategory_crop.py [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+write_image_path = '/home/tamar/DBs/Reccelite/CroppedDB/croppedImgs_1_2_3_4_5_Th06_rare/'
```
[the remark this answers is not in the document]
Comment (TT-happy-work) [reply offline_dataAugmentation_rareCategory_crop.py]: ok

**Pull request merged**